File: 3553.minimum-weighted-subgraph-with-the-required-paths-ii.java
#
# @lc app=leetcode id=3553 lang=java
#
# [3553] Minimum Weighted Subgraph With the Required Paths II
#
# @lc code=start
import java.util.*;
class Solution {
    int LOG = 17; // since n <= 1e5, log2(1e5) ~ 17
    int n;
    List<int[]>[] tree;
    int[] depth, parent, tin, tout;
    int[][] up;
    long[] weightSum;
    int timer = 0;
    void dfs(int u, int p, int d, long w) {
        depth[u] = d;
        tin[u] = timer++;
        parent[u] = p;
        weightSum[u] = w;
        up[u][0] = p == -1 ? u : p;
        for (int i = 1; i < LOG; ++i) {
            up[u][i] = up[up[u][i-1]][i-1];
        }
        for (int[] nei : tree[u]) {
            int v = nei[0], wt = nei[1];
            if (v != p) {
                dfs(v, u, d+1, w + wt);
            }
        }
        tout[u] = timer++;
    }
    boolean isAncestor(int u, int v) {
        return tin[u] <= tin[v] && tout[u] >= tout[v];
    }
    int lca(int u, int v) {
        if (isAncestor(u, v)) return u;
        if (isAncestor(v, u)) return v;
        for (int i = LOG - 1; i >= 0; --i) {
            if (!isAncestor(up[u][i], v)) {
                u = up[u][i];
            }
        }
        return up[u][0];
    }
    public int[] minimumWeight(int[][] edges, int[][] queries) {
        n = edges.length + 1;
        tree = new ArrayList[n];
        for (int i = 0; i < n; ++i) tree[i] = new ArrayList<>();
        for (int[] e : edges) {
            int u = e[0], v = e[1], w = e[2];
            tree[u].add(new int[]{v, w});
            tree[v].add(new int[]{u, w});
        }
        depth = new int[n];
        parent = new int[n];
        tin = new int[n];
        tout = new int[n];
        up = new int[n][LOG];
        weightSum = new long[n];
        dfs(0, -1, 0, 0L);
        int q = queries.length;
        int[] res = new int[q];
        for (int i = 0; i < q; ++i) {
            int a = queries[i][0], b = queries[i][1], c = queries[i][2];
            int l1 = lca(a, c);
            int l2 = lca(b, c);
            int l3 = lca(a, b);
            // The union of paths from a to c and b to c is: sum(a to c) + sum(b to c) - sum(lca(a, b) to c)
            long ans = weightSum[a] + weightSum[b] - weightSum[l3] + weightSum[c] - weightSum[l1] - weightSum[l2] + weightSum[lca(l1, l2)];
            res[i] = (int) ans;
        }
        return res;
    }
}
# @lc code=end

File: 3553.minimum-weighted-subgraph-with-the-required-paths-ii.go
#
# @lc app=leetcode id=3553 lang=golang
#
# [3553] Minimum Weighted Subgraph With the Required Paths II
#
# @lc code=start
func minimumWeight(edges [][]int, queries [][]int) []int {
    n := len(edges) + 1
    tree := make([][][2]int, n)
    for _, e := range edges {
        u, v, w := e[0], e[1], e[2]
        tree[u] = append(tree[u], [2]int{v, w})
        tree[v] = append(tree[v], [2]int{u, w})
    }
    const LOG = 17 // since n <= 1e5, log2(1e5) < 17
    parent := make([][]int, n)
    up := make([][]int, n)
    depth := make([]int, n)
    sum := make([]int, n)
    for i := range parent {
        parent[i] = make([]int, LOG)
        up[i] = make([]int, LOG)
        for j := 0; j < LOG; j++ {
            parent[i][j] = -1
        }
    }
    var dfs func(u, p, d, s int)
    dfs = func(u, p, d, s int) {
        parent[u][0] = p
        depth[u] = d
        sum[u] = s
        for _, e := range tree[u] {
            v, w := e[0], e[1]
            if v != p {
                dfs(v, u, d+1, s+w)
            }
        }
    }
    dfs(0, -1, 0, 0)
    for j := 1; j < LOG; j++ {
        for i := 0; i < n; i++ {
            if parent[i][j-1] != -1 {
                parent[i][j] = parent[parent[i][j-1]][j-1]
            }
        }
    }
    lca := func(u, v int) int {
        if depth[u] < depth[v] {
            u, v = v, u
        }
        for k := LOG-1; k >= 0; k-- {
            if parent[u][k] != -1 && depth[parent[u][k]] >= depth[v] {
                u = parent[u][k]
            }
        }
        if u == v {
            return u
        }
        for k := LOG-1; k >= 0; k-- {
            if parent[u][k] != -1 && parent[u][k] != parent[v][k] {
                u = parent[u][k]
                v = parent[v][k]
            }
        }
        return parent[u][0]
    }
    minSubtreeLCA := func(a, b, c int) int {
        x := lca(a, b)
        y := lca(b, c)
        z := lca(a, c)
        // The LCA of all three nodes is the 'lowest' one among x, y, z
        // Since tree, they are on the same path
        if depth[x] < depth[y] {
            x = y
        }
        if depth[x] < depth[z] {
            x = z
        }
        return x
    }
    ans := make([]int, len(queries))
    for i, q := range queries {
        a, b, c := q[0], q[1], q[2]
        u := minSubtreeLCA(a, b, c)
        total := sum[a] + sum[b] + sum[c] - 2*sum[u]
        // Additional check (domain-agnostic): ensure that all nodes are distinct and subtree covers all required nodes
        // This is always true by construction in a tree, but would be checked if generalizing
        ans[i] = total
    }
    return ans
}
# @lc code=end